### ISSUE TITLE
finetuning sparsified LLaMa

### ISSUE THREAD

**Issue by kiucho** [IST-DASLab/sparsegpt#17]
hello. First of all, thank you for sharing your great research.

I am trying to `fine-tune` the LLaMa-7B model for a specific task after sparsifications.

Is it possible to simply use the saved path via `--save` option?

Thank you.

**Comment by efrantar** [comment]
Hi, `--save` saves a HuggingFace checkpoint of the sparse model where sparse weights are exactly 0. In principle, you should be able to use this with an appropriate finetuning script, however if you want to keep the model sparse, you have to make sure that the exactly 0 weights remain 0. A simple way to accomplish this is to store the mask at the beginning (e.g. via `p == 0` for each parameter in the saved checkpoint) and then zero out the corresponding weights directly after each gradient update (i.e., `optimizer.step()`).

**Comment by kiucho** [comment]
Thanks for your reply!
Let me ask you one more question.

Below is the code that outputs the number of parameters for the dense and pruned models.
![image](https://github.com/IST-DASLab/sparsegpt/assets/91112750/fbd15ee2-87d9-407c-ad52-465ec80dccdf)

1. If their number of parameters are exactly same, is there no advantage of model size?

2. When I set the sparsity to 0.5, 50% of all parameters are set to zero, but it seems that the 50% of parameters that are set to zero are still involved in the multiplication and addition operations during the forward pass. So is there a computational cost benefit to the pruned model?

Thank you.

**Comment by moonlightian** [comment]
> 

Sparsity is one kind of unstructured prune method and it would not change the size of models. 
As for the second question, I am interested too. It seems that computational cost benefit is not optimal but the model can be speedup by CUTLASS.
Maybe not correct, looking forward to author's reply

**Comment by SSshuishui** [comment]
Maybe can use some optimized storage methods to save the sparsity model?  Otherwise, saving the large model will become a problem such as the 175B model.

**Comment by kiucho** [comment]
Now, speedup with CUTLASS is available with Pytorch 2.1. but storage issue is unlikely to be resolved until something else comes along.

Closing this issue.
